common/src/main/java/tachyon/MasterClientBase.java

```diff
@@ -0,0 +1,190 @@
+   * internal state to finish the connection process.
+   */
+  protected abstract void afterConnect();
+
+  /**
+   * This called after the connection is disconnected. Here, implementations should clean up any
+   * additional state created for the connection.
+   */
+  protected abstract void afterDisconnect();
+
+  /**
+   * Connect with the master; an exception is thrown if this fails.
+   *
+   * @throws IOException
+   */
+  public synchronized void connect() throws IOException {
+    if (mConnected) {
+      return;
+    }
+    disconnect();
+
+    if (mIsClosed) {
+      throw new IOException("Client is closed, will not try to connect");
+    }
+
+    Exception lastException = null;
+    int maxConnectsTry = mTachyonConf.getInt(Constants.MASTER_RETRY_COUNT);
+    RetryPolicy retry = new ExponentialBackoffRetry(50, Constants.SECOND_MS, maxConnectsTry);
+    do {
+      mMasterAddress = getMasterAddress();
+
+      LOG.info("Tachyon client (version " + Version.VERSION + ") is trying to connect with "
+          + getServiceName() + " master @ " + mMasterAddress);
+
+      TProtocol binaryProtocol = new TBinaryProtocol(new TFramedTransport(
+          new TSocket(NetworkAddressUtils.getFqdnHost(mMasterAddress), mMasterAddress.getPort())));
+      mProtocol = new TMultiplexedProtocol(binaryProtocol, getServiceName());
+      try {
+        mProtocol.getTransport().open();
+      } catch (TTransportException e) {
+        lastException = e;
+        LOG.error("Failed to connect (" + retry.getRetryCount() + ") to " + getServiceName()
+            + " master @ " + mMasterAddress + " : " + e.getMessage());
+        continue;
+      }
+
+      LOG.info("Client registered with " + getServiceName() + " master @ " + mMasterAddress);
+      mConnected = true;
+      afterConnect();
+      return;
+    } while (retry.attemptRetry() && !mIsClosed);
+
+    // Reaching here indicates that we did not successfully connect.
+    throw new IOException("Failed to connect to " + getServiceName() + " master @ " + mMasterAddress
+        + " after " + (retry.getRetryCount()) + " attempts", lastException);
+  }
+
+  /**
+   * Close the connection with the Tachyon Master and do the necessary cleanup. It should be used if
+   * the client has not connected with the master for a while, for example.
+   */
+  public synchronized void disconnect() {
+    if (mConnected) {
+      LOG.debug("Disconnecting from the " + getServiceName() + " master {}", mMasterAddress);
+      mConnected = false;
+    }
+    try {
+      if (mProtocol != null) {
+        mProtocol.getTransport().close();
+      }
+    } finally {
+      afterDisconnect();
+    }
+  }
+
+  /**
+   * Returns the connected status of the client.
+   *
+   * @return true if this client is connected to the master
+   */
+  public synchronized boolean isConnected() {
+    return mConnected;
+  }
+
+  /**
+   * Close the connection with the master permanently. This instance should be reused after closing.
```

Comment:
Permanently and then reuse seems kind of strange, we are using the close once and it is forever closed semantic right?